developer-workflow/extension-modules.rst

```diff
@@ -5,13 +5,571 @@
 Standard library extension modules
 ==================================
 
-In this section, we could explain how to write a CPython extension with the C language, but the topic can take a complete book.
+In this section, we are interested in extending the CPython project with
+an :term:`extension module`. We will not explain how to write the module
+in C but rather explain how to configure the project and make it compile.
 
-For this reason, we prefer to give you some links where you can read a very good documentation.
+Some modules in the standard library, such as :mod:`datetime` or :mod:`pickle`,
+have identical implementations in C and Python; the C implementation, when
+available, is expected to improve performance (such extension modules are
+commonly referred to as *accelerator modules*).
 
-Read the following references:
+Other modules mainly implemented in Python may import a C helper extension
+providing implementation details (for instance, the :mod:`csv` module uses
+the internal :mod:`!_csv` module defined in :cpy-file:`Modules/_csv.c`).
+
+For writing an extension module in C, we prefer to give you some links
+where you can read good documentation:
 
 * https://docs.python.org/dev/c-api/
 * https://docs.python.org/dev/extending/
 * :pep:`399`
 * https://pythonextensionpatterns.readthedocs.io/en/latest/
+
+Classifying Extension Modules
+=============================
+
+Extension modules can be classified into two categories:
+
+* A *built-in* extension module is a module built and shipped with
+  the Python interpreter. A built-in module is *statically* linked
+  into the interpreter, thereby lacking a :attr:`__file__` attribute.
+
+  .. seealso:: :data:`sys.builtin_module_names` --- names of built-in modules.
+
+* A *shared* (or *dynamic*) extension module is built as a shared library
+  (``.so`` or ``.dll`` file) and is *dynamically* linked into the interpreter.
+
+  In particular, the module's :attr:`__file__` attribute contains the path
+  to the ``.so`` or ``.dll`` file.
+
+.. note::
+
+   Informally, built-in extension modules can be regarded as *required*
+   while shared extension modules are *optional* in the sense that they
+   might be supplied, overridden or disabled externally.
+
+   Usually, accelerator modules are built as *shared* extension modules,
+   especially if they already have a pure Python implementation.
+
+According to :pep:`399`, *new* extension modules MUST provide a working and
+tested pure Python implementation, unless a special dispensation is given.
+Please ask the :github:`Steering Council <python/steering-council>` if such
+dispensation is needed.
+
+While new built-in extension modules could be considered exceptions to that
+rule, shared extension modules must provide a Python implementation, not only
+because of the guidelines, but also as a fallback if the corresponding shared
+library is not be present on the system.
```

Comment:
I'd remove this paragraph, and leave the reasoning to the PEP.
The PEP says:
> Typically the exemption will be granted only when a module wraps a specific C-based library (e.g., [sqlite3](http://docs.python.org/py3k/library/sqlite3.html)).

That still holds. If `sqlite3` was being considered today, it probably wouldn't be added to stdlib at all, but if it would, it would be a shared extension without a Python implementation.

Reply:
> [...] but if it would, it would be a shared extension without a Python implementation.

... which _is_ the case today; there is no Python implementation.

Reply:
Yup.
Compare to something like `binascii`, which also doesn't have a Python implementation, but would absolutely need one if it was added today.